File: solutions/21. Merge Two Sorted Lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode head1, ListNode head2) {
            ListNode t1 = head1;
            ListNode t2 = head2;
​
            ListNode dummy = new ListNode(-100);
            ListNode temp = dummy;
​
            while(t1 != null && t2 != null){
                if(t1.val < t2.val) {
                    temp.next = t1;
                    t1 = t1.next;
                } else {
                    temp.next = t2;
                    t2 = t2.next;    
                }
                temp = temp.next;
            }
            if(t1 != null) {
                temp.next = t1;
                // t1 = t1.next;
            }
​
            if(t2 != null) {
                temp.next = t2;
                // t2 = t2.next;
            }
​
            return dummy.next;
    }
}

File: ..solutions/21. Merge Two Sorted Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        ListNode fake(-1);
        ListNode* last = &fake;
        
        while(l1 != NULL && l2 != NULL){
            if(l1->val < l2->val){
                last->next = l1;
                last = l1;
                l1 = l1->next;
            }else{
                last->next = l2;
                last = l2;
                l2 = l2->next;
            }
        }
        
        if(l1 != NULL){
            last->next = l1;
        }
        if(l2 != NULL){
            last->next = l2;
        }
        return fake.next;
    }
};
